### ISSUE TITLE
compile() drops metric names

### ISSUE THREAD

**Issue by t-kalinowski** [rstudio/keras#450]
`compile()` doesn't pass through a dictionary of metrics, and therefor doesn't allow for matching metrics to a specific named layer. 

Specifically, the names are dropped here https://github.com/rstudio/keras/blob/master/R/model.R#L278

here is a minimal example demonstrating the issue:
``` r
library(keras)

input <- layer_input(10)

core <- input %>% layer_dense(10)

output1 <- core %>% layer_dense(10, name = "output1")
output2 <- core %>% layer_dense(10, name = "output2")
output3 <- core %>% layer_dense(10, name = "output3")

model <- keras_model(input, list(
  output1 = output1,
  output2 = output2,
  output3 = output3
))

model %>% compile(
  "rmsprop", loss = 'mse', 
  metrics = list(
    output1 = 'accuracy',
    output2 = 'mse',
    output3 = 'mae'
  ) 
)

model$metrics_names
#>  [1] "loss"                        "output1_loss"               
#>  [3] "output2_loss"                "output3_loss"               
#>  [5] "output1_acc"                 "output1_mean_squared_error" 
#>  [7] "output1_mean_absolute_error" "output2_acc"                
#>  [9] "output2_mean_squared_error"  "output2_mean_absolute_error"
#> [11] "output3_acc"                 "output3_mean_squared_error" 
#> [13] "output3_mean_absolute_error"

# works as intended with the python call:
model$compile(
  "rmsprop", loss = 'mse', 
  metrics = list(
    output1 = 'accuracy',
    output2 = 'mse',
    output3 = 'mae'
  ) 
)
model$metrics_names
#> [1] "loss"                        "output1_loss"               
#> [3] "output2_loss"                "output3_loss"               
#> [5] "output1_acc"                 "output2_mean_squared_error" 
#> [7] "output3_mean_absolute_error"
```

Created on 2018-06-25 by the [reprex package](http://reprex.tidyverse.org) (v0.2.0).

**Comment by jjallaire** [comment]
There is a conflict between our use of the named list (to provide nice names for R custom metric functions) and the Keras use of the named list (to map metrics to model outputs). I think I can overcome this by checking explicitly for the output names and doing the right thing in that case. More later this week.

**Comment by t-kalinowski** [comment]
What do you think about, for custom metrics, providing users with new constructor like:
```R
metric_lambda(..., name = 'pretty_metric_name')
```
(and deprecating the usage of passing `metrics = list(pretty_metric_name = function(...){...})`

**Comment by jjallaire** [comment]
Yes, I think something like that is definitely the answer as I don't think it will be graceful or clean to try to have the named list serve both masters.
